### PR TITLE
Tutor user role

### PR DESCRIPTION
**Pull request by chrisjpurdy**
This adds the `TUTOR` role, and changes the semantics of different endpoints to cope with this additional role type.

[Front end PR](https://github.com/isaacphysics/isaac-react-app/pull/675)

Tutor welcome email can be found in the content repo(s) as: `email_templates/tutor_welcome.json` (needs to be copyedited first)

---

**Pull Request Check List**
- [x] Unit Tests & Regression Tests Added (Optional)
- [x] Removed Unnecessary Logs/System.Outs/Comments/TODOs
- [x] Added enough Logging to monitor expected behaviour change
- [x] Security - Data Exposure - PII is not stored or sent unencrypted
- [ ] Security - Data Exposure - Test any altered or created endpoints using [swagger](http://localhost:8080/isaac-api/api-docs/)
- [x] Security - Access Control - Check authorisation on every new endpoint
- [x] Peer-Reviewed
- [x] Email templates copyedited

### CODE REVIEW

File: src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java [hunk 1]
```diff
@@ -110,5 +114,137 @@ TEST_TUTOR_ID, TEST_TUTORS_AB_GROUP_ID, new Date(), DateUtils.addDays(new Date()
...
+        assertEquals(Response.Status.OK.getStatusCode(), previewQuizResponse.getStatus());
+
+        // check the quiz is returned for preview
+        IsaacQuizDTO responseBody = (IsaacQuizDTO) previewQuizResponse.getEntity();
+        assertEquals(QUIZ_INVISIBLE_TO_STUDENTS_QUIZ_ID, responseBody.getId());
+    }
+
+    @Test
+    @Ignore("Skipping as this fails - we should decide what the expectations are here.")
+    public void previewQuizEndpoint_previewInvisibleToStudentQuizAsTutor_fails() throws NoCredentialsAvailableException,
```
Comment (mwtrew) [inline src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:202]: As the comment says, I'm not sure if quizzes where `visibleToStudents: false` should be preview-able by tutors.

Comment (chrisjpurdy) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:202]: I think that's an edge case I missed, cheers for spotting that 🙌 Those definitely shouldn't be preview-able since it goes right against the semantics of the `visibleToStudents` flag.

Comment (chrisjpurdy) [reply src/test/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacadeIT.java:202]: I'll fix this (shouldn't need any front-end since tutors only see the student-only tests anyway, but I'll double check that this is the case)

File: src/main/java/uk/ac/cam/cl/dtg/isaac/api/AssignmentFacade.java [hunk 2]
```diff
@@ -988,11 +988,11 @@ public Response assignGameBoards(@Context final HttpServletRequest request,
...
 
-            // Assert user is allowed to set assignments
-            boolean userIsTeacherOrAbove = isUserTeacherOrAbove(userManager, currentlyLoggedInUser);
+            // Assert user is allowed to set assignments - tutors and above are allowed to do so
+            boolean userIsTutorOrAbove = isUserTutorOrAbove(userManager, currentlyLoggedInUser);
             boolean userIsStaff = isUserStaff(userManager, currentlyLoggedInUser);
-            if (!userIsTeacherOrAbove) {
-                return new SegueErrorResponse(Status.FORBIDDEN, "You need a teacher account to create groups and set assignments!").toResponse();
+            if (!userIsTutorOrAbove) {
+                return new SegueErrorResponse(Status.FORBIDDEN, "You need at least a tutor account to create groups and set assignments!").toResponse();
```
Comment (jsharkey13) [inline src/main/java/uk/ac/cam/cl/dtg/isaac/api/AssignmentFacade.java:995]: Maybe nobody sees this message, but "You need a tutor or teacher account ..." is better; staff don't need this message, and users may not know what "at least" means here?

File: src/main/java/uk/ac/cam/cl/dtg/isaac/api/IsaacController.java [hunk 3]
```diff
@@ -371,9 +371,10 @@ public final Response getDocumentByPath(@Context final Request request, @Context
         try {
 
             RegisteredUserDTO currentlyLoggedInUser = userManager.getCurrentRegisteredUser(httpServletRequest);
-            if (!isUserTeacherOrAbove(userManager, currentlyLoggedInUser)) {
+            // Tutors and above should be able to access these documents as teaching material
+            if (!isUserTutorOrAbove(userManager, currentlyLoggedInUser)) {
                 return new SegueErrorResponse(Status.FORBIDDEN,
-                        "You must have a teacher account to access these resources.").toResponse();
+                        "You must have at least a tutor account to access these resources.").toResponse();
```
Comment (jsharkey13) [inline src/main/java/uk/ac/cam/cl/dtg/isaac/api/IsaacController.java:377]: Ditto; just "... have a tutor or teacher account ...".

File: src/main/java/uk/ac/cam/cl/dtg/isaac/api/IsaacController.java [hunk 4]
```diff
@@ -515,7 +516,11 @@ public final Response getUserProgressInformation(@Context final HttpServletReque
             }
             userOfInterestSummary = userManager.convertToUserSummaryObject(userOfInterestFull);
 
-            if (associationManager.hasPermission(user, userOfInterestSummary)) {
+            // If user is fetching their own progress, or the user is at least a teacher and has a valid connection
+            // with this user...
+            if (user.getId().equals(userOfInterestSummary.getId())
+                    || isUserTeacherOrAbove(userManager, user) && associationManager.hasPermission(user, userOfInterestSummary)
```
Comment (jsharkey13) [inline src/main/java/uk/ac/cam/cl/dtg/isaac/api/IsaacController.java:522]: Is this to prevent tutors from seeing My Progress pages? It's not obvious that this is what it is doing, nor why?

File: src/main/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacade.java [hunk 5]
```diff
@@ -339,7 +342,12 @@
             IsaacQuizDTO quiz = this.quizManager.findQuiz(quizId);
 
             // Check this user is actually allowed to preview this quiz:
-            if (null != quiz.getHiddenFromRoles() && quiz.getHiddenFromRoles().contains(user.getRole().name())) {
+            if (!quiz.getVisibleToStudents() && user.getRole() == Role.TUTOR
```
[the remark this answers is not in the document]
Comment (jsharkey13) [reply src/main/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacade.java:345]: This is correct. `getVisibleToStudents` may not be set, and may be incorrect if it is set. Only use `getHiddenFromRoles`.

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java [hunk 6]
```diff
@@ -269,6 +269,9 @@ public synchronized Response modifyUsersRole(@Context final HttpServletRequest r
                 RegisteredUserDTO user = this.userManager.getUserDTOById(userid);
                 Role oldRole = user.getRole();
                 this.userManager.updateUserRole(userid, requestedRole);
+                log.info("ADMIN user " + requestingUser.getEmail() + " has modified the role of "
+                        + user.getEmail() + "[" + user.getId() + "]" + " to "
+                        + user.getRole());
```
Comment (jsharkey13) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/AdminFacade.java:274]: Please use `String.format()`?

File: src/main/java/uk/ac/cam/cl/dtg/isaac/api/IsaacController.java (hunk 4 above)
Comment (jsharkey13) [reply src/main/java/uk/ac/cam/cl/dtg/isaac/api/IsaacController.java:522]: There are a few other places where this pattern occurs; it would be nice to wrap this all up in the AssociationManager as a sensibly-named method, to check role and permission together.

However, it raises another issue; currently if your account is demoted from Teacher to Student, you can still access all the connected students' data, because we only check for the existence of a connection. This will break that. That's probably for the best, but then, maybe the existing check should already verify Tutor or above, and we have a new method `hasTeacherPermission` that's slightly stricter?

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManager.java [hunk 7]
```diff
@@ -562,21 +562,9 @@ public Response updateUserObject(final HttpServletRequest request, final HttpSer
...
-                            "You cannot modify admin roles.").toResponse();
-                }
-            }
-
-            // check that the user is allowed to change the role of another user
-            // if that is what they are doing.
-            if (!checkUserRole(currentlyLoggedInUser, Arrays.asList(Role.ADMIN, Role.EVENT_MANAGER))
-                    && userObjectFromClient.getRole() != null
-                    && !userObjectFromClient.getRole().equals(existingUserFromDb.getRole())) {
+            if (null == userObjectFromClient.getRole() || !existingUserFromDb.getRole().equals(userObjectFromClient.getRole())) {
```
Comment (jsharkey13) [inline src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManager.java:565]: This worried me, until I read it more carefully. Could we have a comment like "You cannot modify role using this endpoint" to make it more clear what this is doing?

**Review by jsharkey13 — COMMENTED** [review]
I didn't review the tests at all. A few comments.

File: src/main/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacade.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrisjpurdy) [reply src/main/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacade.java:345]: Why would it be incorrect if set? I'm fine with removing this but only if each tests `hiddenFromRoles` was generated from existing `visibleToStudents` flags when it was introduced?

File: src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManager.java (hunk 7 above)
Comment (chrisjpurdy) [reply src/main/java/uk/ac/cam/cl/dtg/segue/api/managers/UserAccountManager.java:565]: I thought that the message included in the error response immediately afterwards was enough to indicate what the intention of this `if` statement was. Happy to include a comment though 👍

File: src/main/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacade.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (chrisjpurdy) [reply src/main/java/uk/ac/cam/cl/dtg/isaac/api/QuizFacade.java:345]: On reflection I think I remember that this is exactly what happened when `hiddenFromRoles` was introduced - I'll remove it, along with the test

File: src/main/java/uk/ac/cam/cl/dtg/isaac/api/IsaacController.java (hunk 4 above)
Comment (chrisjpurdy) [reply src/main/java/uk/ac/cam/cl/dtg/isaac/api/IsaacController.java:522]: `isUserXOrAbove` and variants aren't available in `UserAssociationManager` since it isn't a façade, so I had to implement this in a slightly messy way, but it works as intended for our current role structure.
